diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
index b9f5106..fbc2afc 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -365,7 +365,7 @@ MediaTime MediaPlayerPrivateGStreamer::playbackPosition() const
         return duration.isInvalid() ? MediaTime::zeroTime() : duration;
     }
 
-    double now = WTF::currentTime();
+    double now(WTF::monotonicallyIncreasingTime());
     // This cachedPosition time-to-live should be slightly lower than the maxTimeupdateEventFrequency
     // in HTMLMediaElement to prevent returning the same cached value in two consecutive updates.
     if (m_lastQuery > -1 && ((now - m_lastQuery) < 0.01) && m_cachedPosition.isValid())
