From 6b93d3a0a171a5162299d72d360505389cc82348 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Enrique=20Oca=C3=B1a=20Gonz=C3=A1lez?= <eocanha@igalia.com>
Date: Tue, 17 Jul 2018 09:10:39 +0000
Subject: [PATCH] [GStreamer] Fix critical error when using hole punch

---
 .../gstreamer/MediaPlayerPrivateGStreamerBase.cpp    | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp
index cd00e682c62..e610b50d5bb 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp
@@ -487,8 +487,16 @@ FloatSize MediaPlayerPrivateGStreamerBase::naturalSize() const
 	// If there's a sample, get the caps from it.
         caps = gst_sample_get_caps(m_sample.get());
     } else {
-	// If there's no sample, try to get the caps from the video sink.
-        GRefPtr<GstPad> videoSinkPad = adoptGRef(gst_element_get_static_pad(m_videoSink.get(), "sink"));
+        // If there's no sample, try to get the caps from the video sink.
+        GRefPtr<GstElement> videoSink = m_videoSink;
+
+        // m_videoSink can be null for auto-plugged sinks, but we can ask PlayBin.
+        if (!videoSink)
+          g_object_get(m_pipeline.get(), "video-sink", &videoSink.outPtr(), nullptr);
+        if (!videoSink)
+          return { };
+
+        GRefPtr<GstPad> videoSinkPad = adoptGRef(gst_element_get_static_pad(videoSink.get(), "sink"));
         if (videoSinkPad)
             caps = gst_pad_get_current_caps(videoSinkPad.get());
     }
-- 
2.17.0

